tests/transforms/test_unitary_to_rot.py

```diff
@@ -335,289 +342,248 @@ def qfunc_with_qubit_unitary(angles):
-
 
 class TestTwoQubitUnitaryDifferentiability:
     """Tests to ensure the transform is fully differentiable in all interfaces
-    when the circuit contains two-qubit QubitUnitary operations."""
+    when the circuit contains two-qubit QubitUnitary operations.
+
+    Note that we are not testing whether we can differentiate w.r.t. the two-qubit
+    unitary itself, but rather that the entire pipeline remains differentiable even
+    when we are decomposing two-qubit unitaries within.
+    """
 
-    @pytest.mark.parametrize("diff_method", diff_methods)
-    @pytest.mark.parametrize("U", test_u4_unitaries)
-    def test_gradient_unitary_to_rot_two_qubit(self, U, diff_method):
+    @pytest.mark.parametrize("diff_method", ["parameter-shift", "backprop"])
+    def test_gradient_unitary_to_rot_two_qubit(self, diff_method):
         """Tests differentiability in autograd interface."""
+        U0 = np.array(test_two_qubit_unitaries[0], requires_grad=False, dtype=np.complex128)
+        U1 = np.array(test_two_qubit_unitaries[1], requires_grad=False, dtype=np.complex128)
 
-        U = np.array(U, requires_grad=True)
+        def two_qubit_decomp_qnode(x, y, z):
+            qml.RX(x, wires=0)
+            qml.RY(y, wires=1)
+            qml.QubitUnitary(U0, wires=[0, 1])
+            qml.RZ(z, wires=2)
+            qml.QubitUnitary(U1, wires=[1, 2])
+            return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1) @ qml.PauliZ(2))
 
-        original_qnode = qml.QNode(
-            qfunc_with_qubit_unitary_two_qubit, device=dev, diff_method=diff_method
-        )
+        x = np.array(0.1, requires_grad=True)
+        y = np.array(0.2, requires_grad=True)
+        z = np.array(0.3, requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=3)
 
-        transformed_qfunc = unitary_to_rot(qfunc_with_qubit_unitary_two_qubit)
+        original_qnode = qml.QNode(two_qubit_decomp_qnode, device=dev, diff_method=diff_method)
+
+        transformed_qfunc = unitary_to_rot(two_qubit_decomp_qnode)
         transformed_qnode = qml.QNode(transformed_qfunc, dev, diff_method=diff_method)
 
-        assert qml.math.allclose(original_qnode(U), transformed_qnode(U))
+        assert qml.math.allclose(original_qnode(x, y, z), transformed_qnode(x, y, z))
+
+        # 3 normal operations + 10 for the first decomp and 2 for the second
+        assert len(transformed_qnode.qtape.operations) == 15
 
-        original_grad = qml.grad(original_qnode)(U)
-        transformed_grad = qml.grad(transformed_qnode)(U)
+        original_grad = qml.grad(original_qnode)(x, y, z)
+        transformed_grad = qml.grad(transformed_qnode)(x, y, z)
 
         assert qml.math.allclose(original_grad, transformed_grad, atol=1e-6)
 
-    @pytest.mark.parametrize("U", test_u4_unitaries)
-    def test_gradient_unitary_to_rot_torch_two_qubit(self, U):
-        """Tests differentiability in autograd interface."""
+    def test_gradient_unitary_to_rot_torch_two_qubit(self):
```

Comment:
wait the gradient tests are also passing?!?!?!

Reply:
The gradients themselves are not with respect the unitaries we are decomposing, but rather the QNode parameters. I did try to get it to actually differentiate w.r.t. the matrix elements but `det`, `eig`, cause issues when trying to do that in certain frameworks. I also don't think there's much of a use case for that - it's fairly common to use a parameter or two to create a single-qubit unitary, but trickier to create a 2-qubit matrix element-by-element based on some parameters (and if you are doing so, good chance you might already know a decomposition for it). But it is also a 'nice to have' for the future.

Reply:
Okay *almost* true. I had to update JAX on my machine to run some more tests, and that one is not passing. Even if I specify the argnums to use only argument 0, it complains that it cannot differentiate w.r.t. arguments 1 and 2, which are the unitaries.